src/borrow_tracker/tree_borrows/tree.rs

```diff
@@ -725,27 +736,480 @@ mod commutation_tests {
+        // with fixed interleaving, which produces all possibilities of interleavings
+        // up to repeated applications of idempotent transitions)
+        #[derive(Clone, Debug)]
+        struct Pattern {
+            x_initial_state: LocationState,
+            y_initial_state: LocationState,
+            y_protected: bool,
+            opaque: Opaque2,
+            y_return: bool,
+        }
+
+        impl Exhaustive for Pattern {
+            fn exhaustive() -> Box<dyn Iterator<Item = Self>> {
+                let mut v = Vec::new();
+                for (x_initial_state, y_initial_state) in
+                    <(LocationState, LocationState)>::exhaustive()
+                {
+                    // We can only do spurious reads for initialized locations anyway.
+                    precondition!(x_initial_state.initialized);
+                    // Moreover we don't create `Active` or `Disabled` directly on reborrow.
+                    precondition!(!x_initial_state.permission.is_disabled());
+                    precondition!(!y_initial_state.permission.is_disabled());
+                    precondition!(!x_initial_state.permission.is_active());
+                    precondition!(!y_initial_state.permission.is_active());
+                    for [y_protected, y_return] in <[bool; 2]>::exhaustive() {
+                        // Restrict options on when y can lose its protector
+                        precondition!(y_protected || !y_return);
+                        for opaque in Opaque2::exhaustive() {
+                            v.push(Pattern {
+                                x_initial_state,
+                                y_initial_state,
+                                y_protected,
+                                opaque,
+                                y_return,
+                            });
                         }
                     }
                 }
+                Box::new(v.into_iter())
             }
         }
+
+        impl fmt::Display for Pattern {
+            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+                write!(f, "retag x ({}, protect); ", self.x_initial_state)?;
+                write!(
+                    f,
+                    "retag y ({}, {}protect); ",
+                    self.y_initial_state,
+                    if self.y_protected { "" } else { "no" }
+                )?;
+                write!(f, "{}", self.opaque)?;
+                if self.y_return {
+                    write!(f, "ret y")?;
+                }
+                Ok(())
+            }
+        }
+
+        impl Pattern {
+            fn execute(&self, spurious_read: bool) -> Result<LocStateProt2, ()> {
+                let state = LocStateProt2 {
+                    x: LocStateProt { state: self.x_initial_state, prot: true },
+                    y: LocStateProt { state: self.y_initial_state, prot: self.y_protected },
+                };
+                // retag y
```

Comment:
The text also says there's a "retag x". Where is that?
And below is a read, not a retag... needs more explanation.